build/config/BUILD.gn

```diff
@@ -84,6 +84,11 @@ config("debug") {
 config("release") {
   defines = [ "NDEBUG" ]
 
+  if (is_mac || is_ios) {
+    # Disable NSAsserts in release builds.
+    defines += [ "NS_BLOCK_ASSERTIONS=1" ]
+  }
```

Comment:
Not really, we don't use foundation assertions.

If the assertion needs to be fatal in all modes, there is `FML_CHECK`. If its a useful signal while developing the engine, there is `FML_DCHECK`. There are vanishingly few cases where the algorithmic complexity of an operation changes based on the either mode so it mostly just to indicate "actionable to end-users" vs "actionable to engine-developers". If its actionable to neither, there is typically no assertion.

The contributing factor for this suppression [seems to be here](https://github.com/flutter/engine/blob/6264763a8282799f812b90c5b34bea352f49726c/shell/platform/darwin/macos/framework/Source/FlutterViewController.mm#L522). I think that code handles the case via an early return.

It depends on whether the macOS/iOS embedder authors expect assertions to be blocked in release modes and whether this is a good assertion to have. At least to my reading, it looks like no one understands how we got into this situation and so we are disabling the assertion. But the ability to reason about the severity of ignoring the assertion and whether ignoring it will cause issues further down the line is perhaps lost now.

Reply:
Agreed with @chinmaygarde. I don't think we'd left them enabled intentionally, but rather I think mostly assumed we were using the FML macros.

To be honest, I thought I had gone through and replaced any I found with FML macros at least in the macOS embedder, but maybe that was just cleaning up Windows assertions (e.g. https://github.com/flutter/engine/pull/38826), and just imagined I'd also got around to it for macOS.

We should probably come to a decision as to whether we want to standardise on one other the other so we don't have a semi-random mix of the two. Personal preference would be to standardise on FML across the board both for consistency and because they provide more indication of intent, but either way is fine.

Reply:
I think standardizing on the macros makes senses, though I think unless we have a lint forbidding `NSAsserts` and friends, they will sneak in via reviewers not being aware its disallowed.  
[NSAssert](https://github.com/search?q=repo%3Aflutter%2Fengine%20NSAssert%20&type=code)
[NSCAssert](https://github.com/search?q=repo%3Aflutter%2Fengine+NSCAssert+&type=code)

I think we should turn this on to match the Xcode default (`ENABLE_NS_ASSERTIONS` set to `NO` in release) of having `NSAssert`s fire in debug but not in-the-wild production apps.  But that wouldn't prevent us from auditing the embedder code to swap to macros.